crates/driver/src/tests/cases/protocol_fees.rs

```diff
@@ -136,55 +192,173 @@ async fn surplus_protocol_fee_sell_order_capped() {
-        solver_fee: Some(10.ether().into_wei()),
+        network_fee: Some(10.ether().into_wei()),
         quote_sell_amount: 50.ether().into_wei(),
         quote_buy_amount: 40.ether().into_wei(),
         executed: 40.ether().into_wei(),
         executed_sell_amount: 50.ether().into_wei(),
         executed_buy_amount: 36.ether().into_wei(),
     };
 
-    protocol_fee_test_case(test_case).await;
+    common_fee_test_case(test_case).await;
 }
 
 #[tokio::test]
 #[ignore]
 async fn volume_protocol_fee_buy_order() {
     let fee_policy = FeePolicy::Volume { factor: 0.5 };
-    let test_case = TestCase {
+    let test_case = CommonTestCase {
         order_side: order::Side::Buy,
         fee_policy,
         order_sell_amount: 50.ether().into_wei(),
-        solver_fee: Some(10.ether().into_wei()),
+        network_fee: Some(10.ether().into_wei()),
         quote_sell_amount: 50.ether().into_wei(),
         quote_buy_amount: 40.ether().into_wei(),
         executed: 40.ether().into_wei(),
         executed_sell_amount: 75.ether().into_wei(),
         executed_buy_amount: 40.ether().into_wei(),
     };
 
-    protocol_fee_test_case(test_case).await;
+    common_fee_test_case(test_case).await;
 }
 
 #[tokio::test]
 #[ignore]
 async fn volume_protocol_fee_sell_order() {
     let fee_policy = FeePolicy::Volume { factor: 0.5 };
-    let test_case = TestCase {
+    let test_case = CommonTestCase {
         order_side: order::Side::Sell,
         fee_policy,
         order_sell_amount: 50.ether().into_wei(),
-        solver_fee: Some(10.ether().into_wei()),
+        network_fee: Some(10.ether().into_wei()),
         quote_sell_amount: 50.ether().into_wei(),
         quote_buy_amount: 40.ether().into_wei(),
         executed: 40.ether().into_wei(),
         executed_sell_amount: 50.ether().into_wei(),
         executed_buy_amount: 20.ether().into_wei(),
     };
 
-    protocol_fee_test_case(test_case).await;
+    common_fee_test_case(test_case).await;
+}
+
+// Price Improvement policy fee tests.
+// Out of market order could be defined as:
+//   (order.sell + order.fee) * quote.buy < (quote.sell + quote.fee) * order.buy
+// In the following tests Limit orders are used only, where order fee is 0. The
+// amount values are adjusted to respect the definition.
+
+#[tokio::test]
+#[ignore]
+async fn price_improvement_fee_buy_out_of_market_order() {
+    let order_sell_amount = 50.ether().into_wei();
+    let order_buy_amount = 40.ether().into_wei();
+    let quote_network_fee = 20.ether().into_wei();
+    let quote = PriceImprovementQuote {
+        sell_amount: order_sell_amount,
+        buy_amount: order_buy_amount,
+        network_fee: quote_network_fee,
+    };
+    let test_case = PriceImprovementTestCase {
+        order_side: order::Side::Buy,
+        policy_factor: 0.5,
+        policy_max_volume_factor: 1.0,
+        quote,
+        order_sell_amount,
+        order_buy_amount,
+        network_fee: Some(1.ether().into_wei()),
```

Comment:
You are still not saying anywhere that this is the little detail that leads to price improvement. This is way to hidden to leave this uncommented.